File: ceramic/ceramic/report/lot_wise_balance/lot_wise_balance.js
// Copyright (c) 2015, Frappe Technologies Pvt. Ltd. and Contributors
// License: GNU General Public License v3. See license.txt

frappe.query_reports["Lot-Wise Balance"] = {
	"filters": [
		{
			"fieldname": "company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"width": "80",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd": 1
		},
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"width": "80",
			"default": frappe.datetime.get_today(),
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"width": "80",
			"default": frappe.datetime.get_today()
		},
		{
			"fieldname": "item_group",
			"label": __("Item Group"),
			"fieldtype": "MultiSelectList",
			"get_data": function (text) {
				//if (!frappe.query_report.item_group) return;
				return frappe.db.get_link_options('Item Group', text)
			},
			"change": function () {
				frappe.query_report.refresh();
			}
		},
		{
			"fieldname": "tile_quality",
			"label": __("Tile Qulaity"),
			"fieldtype": "MultiSelectList",
			"get_data": function (text) {
				//if (!frappe.query_report.item_group) return;
				return frappe.db.get_link_options('Tile Quality', text)
			},
			"change": function () {
				frappe.query_report.refresh();
			}
		},
		{
			"fieldname": "packing_type",
			"label": __("Packing Type"),
			"fieldtype": "MultiSelectList",
			"get_data": function (text) {
				//if (!frappe.query_report.item_group) return;
				return frappe.db.get_link_options('Packing Type', text)
			},
			"change": function () {
				frappe.query_report.refresh();
			}
		},
		{
			"fieldname": "item_code",
			"label": __("Item Code"),
			"fieldtype": "MultiSelectList",
			"get_data": function (text) {
				//if (!frappe.query_report.item_group) return;
				return frappe.db.get_link_options('Item', text, filters = { 'is_item_series': 0, 'has_batch_no': 1 })
			},
			"change": function () {
				frappe.query_report.refresh();
			},

			// "get_query": function() {
			// 	var item_group = frappe.query_report.get_filter_value('item_group')
			// 	if (item_group){
			// 		return {
			// 			doctype: "Item",
			// 			filters: {
			// 				"item_group": item_group,
			// 				"is_item_series": 0
			// 			}
			// 		}
			// 	} else {
			// 		return {
			// 			doctype: "Item",
			// 			filters: {
			// 				"is_item_series": 0
			// 			}
			// 		}
			// 	}
			// }
		},
		{
			"fieldname": "not_in_item_code",
			"label": __("Not in Item Code"),
			"fieldtype": "MultiSelectList",
			"get_data": function (text) {
				//if (!frappe.query_report.item_group) return;
				return frappe.db.get_link_options('Item', text, filters = { 'is_item_series': 0, 'has_batch_no': 1 })
			},
			"change": function () {
				frappe.query_report.refresh();
			},
		},
		{
			"fieldname": "print_with_picked_qty",
			"label": __("Print With Picked Qty"),
			"fieldtype": "Check",
			"options": "Tile Quality",
		},
		{
			"fieldname": "sales_order",
			"label": __("Sales Order"),
			"fieldtype": "Link",
			"options": "Sales Order",
			"get_query": function() {
				var company = frappe.query_report.get_filter_value('company');
				return {
					"doctype": "Sales Order",
					"filters": {
						"company": ['in', company],
						"docstatus": 1,
						"per_delivered": ['!=', 100],
						"status": ['not in', ('Draft', 'Submitted', 'Closed')]
					}
				}
			}
		},
		{
			"fieldname": "warehouse",
			"label": __("Show Warehouse"),
			"fieldtype": "Check"
		},
	]
}


function get_picked_item_details(item_code, batch_no, company, from_date, to_date, bal_qty, total_picked_qty, total_remaining_qty, lot_no) {
	let template = `
		<table class="table table-borderless" style="border: 0 !important; font-size:95%;">
			<tr style="border: 0 !important;">
				<td style="border: 0 !important;"><b>Lot No: </b> {{ data[0]['lot_no'] }}</td>
				<td style="border: 0 !important;"><b>Picked Qty: </b> {{ data[0]['total_picked_qty'] }}</td>
			</tr>
			<tr style="border: 0 !important;">
				<td style="border: 0 !important;"><b>Qty: </b> {{ data[0]['bal_qty'] }}</td>
				<td style="border: 0 !important;"><b>Available Qty: </b> {{ data[0]['total_remaining_qty'] }}</td>
			</tr>
		</table>
		{% if data[0]['customer'] %}
		<table class="table table-bordered" style="margin: 0; font-size:80%;">
			<thead>
				<th>{{ __("Customer") }}</th>
				<th>{{ __("Rank") }}</th>
				<th>{{ __("Sales Order") }}</th>
				<th>{{ __("SO Date") }}</th>
				<th>{{ __("Pick List") }}</th>
				<th>{{ __("% Picked") }}</th>
				<th>{{ __("Picked") }}</th>
				<th>{{ __("Unpick Qty") }}</th>
				<th></th>
			</thead>
			<tbody>
				{% for (let row of data ) { %}
					<tr class="{{ __(row['pick_list_item']) }}">
						<td>{{ __(row['customer']) }}</td>
						<td>{{ __(row['order_rank']) }}</td>
						<td>{{ __(row['sales_order_link']) }}</td>
						<td>{{ __(row['date']) }}</td>
						<td>{{ __(row['pick_list_link']) }}</td>
						<td>{{ __(frappe.format(row['per_picked'], {'fieldtype': 'Percent'})) }}</td>
						<td>{{ __(row['picked_qty']) }}</td>
						<td><input type="float"  min="0" onkeypress="return (event.charCode == 8 || event.charCode == 0) ? null : event.charCode >= 48 && event.charCode <= 57" style="width:30px" id="{{ row['pick_list_item'] }}"></input></td>
						<td><button style="margin-left:5px;border:none;color: #fff; background-color: red; padding: 3px 5px;border-radius: 5px;" type="button" sales-order="{{ __(row['sales_order']) }}" sales-order-item="{{ __(row['sales_order_item']) }}" pick-list="{{ __(row['pick_list']) }}" pick-list-item="{{ __(row['pick_list_item']) }}" onClick=remove_picked_item_lot_wise(this.getAttribute("sales-order"),this.getAttribute("sales-order-item"),this.getAttribute("pick-list"),this.getAttribute("pick-list-item"),document.getElementById("{{ row['pick_list_item'] }}").value)>Unpick</button></td>
					</tr>
				{% } %}
			</tbody>
		</table>
		{% endif %}`;

	// docudocument.getElementById("demo").innerHTML = item_code;

	frappe.call({
		method: "ceramic.api.get_picked_item",
		args: {
			item_code: item_code,
			batch_no: batch_no,
			from_date: from_date,
			to_date: to_date,
			company: company,
			bal_qty: bal_qty,
			total_picked_qty: total_picked_qty,
			total_remaining_qty: total_remaining_qty,
			lot_no: lot_no
		},
		callback: function (r) {
			let message = frappe.template.compile(template)({ 'data': r.message });
			frappe.msgprint({
				message: message,
				title: "Lot-Wise Balance Details : " + item_code,
				wide: true
			});
		}
	})
}

function remove_picked_item_lot_wise(sales_order, sales_order_item, pick_list, pick_list_item, unpick_qty) {
	frappe.call({
		method: "ceramic.ceramic.doc_events.pick_list.unpick_item",
		args: {
			sales_order: sales_order,
			sales_order_item: sales_order_item,
			pick_list: pick_list,
			pick_list_item: pick_list_item,
			unpick_qty:unpick_qty
		},
		callback: function (r) {
			if(r.message){
				$(".modal").modal('hide');
				setTimeout(function(){frappe.msgprint(r.message)}, 500);
			}
			$('.' + pick_list_item).hide()
		}
	})
}

function new_qty_details(item_code,item_group, balance_qty, warehouse, buying_unit_price, batch_no, lot_no, packing_type) {
	let template = `
		<table class="table table-borderless" style="border: 0 !important; font-size:95%;">
			<tr style="border: 0 !important;">
			<td style="border: 0 !important;"><b>Item Group: </b> {{ data['item_group'] }}</td>
			</tr>
			<tr style="border: 0 !important;">
			<td style="border: 0 !important;"><b>Available Qty : </b>{{data['balance_qty']}}</td>
			<td style="border: 0 !important;"><b>Warehouse : </b>{{data['warehouse']}}</td>
			</tr>
			<tr style="border: 0 !important;">
			<td style="border: 0 !important;"><b>Batch : </b>{{data['batch_no']}}</td>
			<td style="border: 0 !important;"><b>Lot : </b>{{data['lot_no']}}</td>
			</tr>
			<tr style="border: 0 !important;">
			<td style="border: 0 !important;"><input type="float" style="width:50px" id="{{ 'new_qty' }}"></input>
			<td style="border: 0 !important;">
			<button style="margin-left:5px;border:0 !important;color: #fff; background-color: blue; padding: 3px 5px;border-radius: 5px;" type="button" warehouse = "{{ __(data['warehouse']) }}" balance_qty = "{{ __(data['balance_qty']) }}" item_code = "{{ __(data['item_code']) }}" buying_unit_price = "{{ __(data['buying_unit_price']) }}" batch_no = "{{ __(data['batch_no']) }}" lot_no = "{{ __(data['lot_no']) }}" packing_type = "{{ __(data['packing_type']) }}" onClick=create_stock_entry(this.getAttribute("warehouse"),this.getAttribute("item_code"),this.getAttribute("balance_qty"),this.getAttribute("buying_unit_price"),document.getElementById("{{ 'new_qty' }}").value,this.getAttribute("batch_no"),this.getAttribute("lot_no"),this.getAttribute("packing_type"))>Create Stock Entry</button>
			</tr>
		</table>`;
		let message = frappe.template.compile(template)({ 'data': {"item_code":item_code,"item_group":item_group,"balance_qty":balance_qty,"warehouse":warehouse,"buying_unit_price":buying_unit_price,"batch_no":batch_no, "lot_no":lot_no, "packing_type":packing_type} });
		frappe.msgprint({
			message: message,
			title: "Item Code : " + item_code,
			wide: true,
		});
}

function create_stock_entry(warehouse,item_code,balance_qty,buying_unit_price,new_qty,batch_no,lot_no,packing_type) {
	if ((new_qty) < 0){
		frappe.throw("Please Don't Enter Negative Qty")
	}

	$(".modal").modal('hide');
	frappe.call({
		method:"ceramic.ceramic.report.lot_wise_balance.lot_wise_balance.create_stock_entry",
		args:
		{
			warehouse:warehouse,
			item_code:item_code,
			balance_qty:balance_qty,
			buying_unit_price:buying_unit_price,
			new_qty:new_qty,
			batch_no:batch_no,
			lot_no:lot_no,
			packing_type:packing_type
		},
		freeze: true,
		freeze_message: "<b>creating stock entry...<b>",
		callback:function(r){
			frappe.msgprint(r.message)
			//setTimeout(function () {frappe.msgprint(r.message)}, 1000)
			// window.location.reload();
			//setTimeout(function () {$(".modal").modal('hide');}, 1000)	 
		}
	})
}

File: ceramic/ceramic/report/lot_wise_balance/lot_wise_balance.py
# Copyright (c) 2015, Frappe Technologies Pvt. Ltd. and Contributors
# License: GNU General Public License v3. See license.txt

from __future__ import unicode_literals

import frappe
from frappe import _
from frappe.utils import cint, flt, getdate


def execute(filters=None):
	if not filters: filters = {}

	float_precision = cint(frappe.db.get_default("float_precision")) or 3

	columns = get_columns(filters)
	item_pick_map = get_picked_qty(filters,float_precision)
	iwb_map = get_item_warehouse_batch_map(filters, float_precision)
	conditions = ''
	if filters.get("to_date"):
		conditions += " AND pl.posting_date <= '%s'" % filters["to_date"]
	
	data = []
	for item in iwb_map:
		if not filters.get("item") or filters.get("item") == item:
			for batch in sorted(iwb_map[item]):
				qty_dict = iwb_map[item][batch]
				try:
					picked_qty = item_pick_map[item][batch].pickedqty
				except KeyError:
					picked_qty = 0.0
				# frappe.db.sql(f"""
				# SELECT sum(pli.qty - (pli.wastage_qty + pli.delivered_qty)) FROM `tabPick List Item` as pli JOIN `tabPick List` as pl on pli.parent = pl.name 
				# WHERE pli.item_code = '{item}' AND pli.batch_no='{batch}' and pl.docstatus = 1 {conditions} AND pl.company = '{filters.get('company')}'
				# """)[0][0] or 0.0
				detail_button = """
					<button style='margin-left:5px;border:none;color: #fff; background-color: #5e64ff; padding: 3px 5px;border-radius: 5px;' 
						type='button' item-code='{}' batch-no='{}' company='{}' from-date='{}' to-date='{}' bal-qty='{}', total-picked-qty = '{}' total-remaining-qty='{}' lot-no='{}'
						onClick='get_picked_item_details(this.getAttribute("item-code"), this.getAttribute("batch-no"), this.getAttribute("company"), this.getAttribute("from-date"), this.getAttribute("to-date"), this.getAttribute("bal-qty"), this.getAttribute("total-picked-qty"), this.getAttribute("total-remaining-qty"), this.getAttribute("lot-no"))'>View</button>""".format(item, batch, filters.get('company'), filters.get('from_date'), filters.get('to_date'), flt(qty_dict.bal_qty, float_precision), picked_qty, flt(qty_dict.bal_qty, float_precision) - picked_qty, qty_dict.lot_no)
				# # frappe.throw(str(detail_button))
				if qty_dict.opening_qty or qty_dict.in_qty or qty_dict.out_qty or qty_dict.bal_qty:
					data.append(sub_data = {
						'item_code': item,
						'lot_no': qty_dict.lot_no,
						'packing_type': qty_dict.packing_type,
						'balance_qty': flt(qty_dict.bal_qty, float_precision),
						'picked_qty': picked_qty,
						'remaining_qty': flt(qty_dict.bal_qty, float_precision) - picked_qty,
						'picked_detail': detail_button,
						'opening_qty': flt(qty_dict.opening_qty, float_precision),
						'in_qty': flt(qty_dict.in_qty, float_precision),
						'out_qty': flt(qty_dict.out_qty, float_precision), 
						'batch_no': batch,
						'item_group': qty_dict.item_group,
						'tile_quality': qty_dict.tile_quality,
						'item_design': qty_dict.item_design,
						'image': qty_dict.image
					})

	return columns, data

def get_columns(filters):
	"""return columns based on filters"""

	columns = [
		{
			"label": _("Item Code"),
			"fieldname": "item_code",
			"fieldtype": "link",
			"options": "Item",
			"width": 180
		},
		{
			"label": _("Lot No"),
			"fieldname": "lot_no",
			"fieldtype": "Data",
			"width": 80
		},
		{
			"label": _("Packing Type"),
			"fieldname": "packing_type",
			"fieldtype": "link",
			"options": "Packing Type",			
			"width": 80
		},		
		{
			"label": _("Balance Qty"),
			"fieldname": "balance_qty",
			"fieldtype": "Float",
			"width": 80
		},
		{
			"label": _("Picked Qty"),
			"fieldname": "picked_qty",
			"fieldtype": "Float",
			"width": 80
		},
		{
			"label": _("Remaining Qty"),
			"fieldname": "remaining_qty",
			"fieldtype": "Float",
			"width": 80
		},
		{
			"label": _("Details"),
			"fieldname": "picked_detail",
			"fieldtype": "Data",
			"width": 70
		}
	]

	if filters.get('sales_order'):
		columns += [
		{
			"label": _("Sales Order Picked Qty"),
			"fieldname": "so_picked_qty",
			"fieldtype": "Float",
			"width": 70,
			"default": 0
		}
	]

	columns += [
		{
			"label": _("Batch"),
			"fieldname": "batch_no",
			"fieldtype": "Link",
			"options": "Batch",
			"width": 100
		},
		{
			"label": _("Item Group"),
			"fieldname": "item_group",
			"fieldtype": "Link",
			"options": "Item Group",
			"width": 180
		},
		{
			"label": _("Quality"),
			"fieldname": "tile_quality",
			"fieldtype": "Data",
			"width": 70
		},
		{
			"label": _("Item Design"),
			"fieldname": "item_design",
			"fieldtype": "Data",
			"width": 100
		},
		{
			"label": _("Opening Qty"),
			"fieldname": "opening_qty",
			"fieldtype": "Float",
			"width": 80
		},
		{
			"label": _("In Qty"),
			"fieldname": "in_qty",
			"fieldtype": "Float",
			"width": 80
		},
		{
			"label": _("Out Qty"),
			"fieldname": "out_qty",
			"fieldtype": "Float",
			"width": 80
		},
		{
			"label": _("image"),
			"fieldname": "image",
			"fieldtype": "data",
			"width": 80
		}
	]

	return columns


def get_conditions(filters):
	conditions = ""
	if not filters.get("from_date"):
		frappe.throw(_("'From Date' is required"))

	if filters.get("to_date"):
		conditions += " and sle.posting_date <= '%s'" % filters["to_date"]
	else:
		frappe.throw(_("'To Date' is required"))
	
	if filters.get("item_group"):
		group_placeholder = ', '.join(f"'{i}'" for i in filters["item_group"])
		conditions += " and i.item_group in (%s)" % group_placeholder
	
	if filters.get("item_code"):
		item_placeholder= ', '.join(f"'{i}'" for i in filters["item_code"])
		conditions += " and sle.item_code in (%s)" % item_placeholder

	if filters.get("not_in_item_code"):
		not_in_item_placeholder= ', '.join(f"'{i}'" for i in filters["not_in_item_code"])
		conditions += " and sle.item_code not in (%s)" % not_in_item_placeholder
	
	if filters.get("tile_quality"):
		tile_quality_placeholder = ', '.join(f"'{i}'" for i in filters["tile_quality"])
		conditions += " and i.tile_quality in (%s)" % tile_quality_placeholder
	
	if filters.get("packing_type"):
		packing_type_placeholder = ', '.join(f"'{i}'" for i in filters["packing_type"])
		conditions += " and batch.packing_type in (%s)" % packing_type_placeholder

	if filters.get("company"):
		conditions += " and sle.company = '%s'" % filters["company"]
	
	if filters.get("sales_order"):
		so_doc = frappe.get_doc("Sales Order", filters.sales_order)
		so_item_list = [row.item_code for row in so_doc.items]

		so_item_list_placeholder = ', '.join(f"'{i}'" for i in so_item_list)
		conditions += " and sle.item_code in (%s)" % so_item_list_placeholder

	return conditions


# get all details
def get_stock_ledger_entries(filters):
	conditions = get_conditions(filters)
	return frappe.db.sql("""
		select sle.item_code, i.item_group, i.tile_quality, i.item_design, i.image as image, batch.lot_no, batch.packing_type, sle.batch_no, sle.posting_date, sum(actual_qty) as actual_qty
		from `tabStock Ledger Entry` as sle 
		JOIN `tabItem` as i on i.item_code = sle.item_code
		JOIN `tabBatch` as batch on batch.name = sle.batch_no
		where sle.docstatus < 2 and ifnull(sle.batch_no, '') != '' %s
		group by voucher_no, batch_no, item_code
		order by i.item_group, sle.item_code""" %
		conditions, as_dict=1)


def get_item_warehouse_batch_map(filters, float_precision):
	sle = get_stock_ledger_entries(filters)
	iwb_map = {}

	from_date = getdate(filters["from_date"])
	to_date = getdate(filters["to_date"])
	for d in sle:
		iwb_map.setdefault(d.item_code, {})\
			.setdefault(d.batch_no, frappe._dict({
				"opening_qty": 0.0, "in_qty": 0.0, "out_qty": 0.0, "bal_qty": 0.0
			}))
		qty_dict = iwb_map[d.item_code][d.batch_no]
		if d.posting_date < from_date:
			qty_dict.opening_qty = flt(qty_dict.opening_qty, float_precision) \
				+ flt(d.actual_qty, float_precision)
		elif d.posting_date >= from_date and d.posting_date <= to_date:
			if flt(d.actual_qty) > 0:
				qty_dict.in_qty = flt(qty_dict.in_qty, float_precision) + flt(d.actual_qty, float_precision)
			else:
				qty_dict.out_qty = flt(qty_dict.out_qty, float_precision) \
					+ abs(flt(d.actual_qty, float_precision))

		qty_dict.bal_qty = flt(qty_dict.bal_qty, float_precision) + flt(d.actual_qty, float_precision)
		qty_dict.lot_no = d.lot_no
		qty_dict.item_group = d.item_group
		qty_dict.tile_quality = d.tile_quality
		qty_dict.item_design = d.item_design
		qty_dict.packing_type = d.packing_type
		qty_dict.image = d.image
	return iwb_map


def get_picked_qty(filters,float_precision):
	picked = get_picked_items(filters)
	picked_map = {}
	for d in picked:
		picked_map.setdefault(d.item_code, {})\
			.setdefault(d.batch_no, frappe._dict({
				"pickedqty": 0.0
			}))
		picked_dict = picked_map[d.item_code][d.batch_no]
		picked_dict.pickedqty = flt(picked_dict.pickedqty, float_precision) + flt(d.pickedqty, float_precision)

	return picked_map

def get_picked_items(filters):
	conditions = get_picked_conditions(filters)
	return frappe.db.sql(f"""
				SELECT pli.item_code, pli.batch_no, (pli.qty - (pli.wastage_qty + pli.delivered_qty)) as pickedqty
				FROM `tabPick List Item` as pli 
				JOIN `tabPick List` as pl on pli.parent = pl.name 
				JOIN `tabItem` as i on i.item_code = pli.item_code
				WHERE pl.docstatus = 1 {conditions}
				""", as_dict=1)
	
def get_picked_conditions(filters):
	conditions = ""

	if filters.get("to_date"):
		conditions += " and pl.posting_date <= '%s'" % filters["to_date"]
	else:
		frappe.throw(_("'To Date' is required"))
	
	if filters.get("item_group"):
		group_placeholder= ', '.join(f"'{i}'" for i in filters["item_group"])
		conditions += " and pli.item_group in (%s)" % group_placeholder
	
	if filters.get("item_code"):
		item_placeholder= ', '.join(f"'{i}'" for i in filters["item_code"])
		conditions += " and pli.item_code in (%s)" % item_placeholder
	
	if filters.get("not_in_item_code"):
		not_in_item_placeholder= ', '.join(f"'{i}'" for i in filters["not_in_item_code"])
		conditions += " and pli.item_code not in (%s)" % not_in_item_placeholder
	
	
	if filters.get("tile_quality"):
		tile_quality_placeholder = ', '.join(f"'{i}'" for i in filters["tile_quality"])
		conditions += " and i.tile_quality in (%s)" % tile_quality_placeholder
	
	if filters.get("company"):
		conditions += " and pl.company = '%s'" % filters["company"]
	
	if filters.get("sales_order"):
		so_doc = frappe.get_doc("Sales Order", filters.sales_order)
		so_item_list = [row.item_code for row in so_doc.items]

		so_item_list_placeholder = ', '.join(f"'{i}'" for i in so_item_list)
		
		conditions += " and i.item_code in (%s)" % so_item_list_placeholder

	return conditions